File: Problem114/Solution.go
package main

type TreeNode struct {
	Val int
	LeftTreeNode
	RightTreeNode
}

func flatten(root *TreeNode) {
	if root == nil || (root.Left == nil && root.Right == nil) {
		return
	}
	flatten(root.Left)
	flatten(root.Right)
	currRight := root.Right
	root.Right = root.Left
	root.Left = nil
	for root.Right != nil {
		root = root.Right
	}
	root.Right = currRight
}

File: Problem114/Solution.java

public class Solution
{
    public void flatten(TreeNode root)
    {
        if (root == null || (root.left == null && root.right == null))
        {
            return;
        }
        flatten(root.left);
        flatten(root.right);
        TreeNode currRight = root.right;
        root.right = root.left;
        root.left = null;
        while (root.right != null)
        {
            root = root.right;
        }
        root.right = currRight;
    }
}